url29ae058134_yuriy_chumak_jatha_tgz-pJ8-org_jatha_LispJ8/src/org/jatha/read/LispParser.java

```diff
@@ -0,0 +1,1065 @@
+      catch (IOException e)
+      { System.err.println("\n *** I/O error while reading character token."); }
+      if (intCh < 0)  { throw new EOFException("Premature end of LISP input."); }
+      else
+        ch = (char) intCh;
+
+      while (! isTerminator(ch))
+      {
+        token.append(ch);
+        try {
+          intCh = inputReader.read();
+        } catch (IOException e) {
+          System.err.println("\n *** I/O error while reading a #' expression." + e);
+        }
+        if (intCh < 0)
+          throw new EOFException("Premature end of LISP input.");
+        else
+          ch = (char) intCh;
+      }
+
+      try {
+        inputReader.unread(ch);
+      } catch (IOException e) {
+        System.err.println("\n *** I/O error while reading a #' expression." + e);
+        return tokenToLispValue(token.toString());
+      }
+
+      if(token.length()>1) {
+          final String tok = token.toString();
+          if(tok.equalsIgnoreCase("SPACE")) {
+              ch = ' ';
+          } else if(tok.equalsIgnoreCase("NEWLINE")) {
+              ch = '\n';
+          } else {
+              ch = 0;
+          }
+      } else {
+          ch = token.charAt(0);
+      }
+
+      return new StandardLispCharacter(ch);
+    }
+
+    // #< usually starts a structure
+    else if (isLeftAngleBracket(ch))
+    {
+      System.err.println("\n *** parser can't read structures yet.");
+      while (! isRightAngleBracket(ch))
+        try {
+          intCh = inputReader.read();
+          ch = (char)intCh;
+        } catch (IOException e) {
+          System.err.println("\n *** I/O error while reading a structure.");
+        }
+
+      if (intCh < 0)
+      {
+        throw new EOFException("Premature end of LISP input.");
+      }
+
+      else
+        ch = (char) intCh;
+
+      return NIL;
+    }
+
+    // -- #| ... |# is a block comment
+    else if (isOrBar(ch))
+    {
+      try {
+        boolean done = false;
+        boolean terminating = false;
+        while (! done)
+        {
+          intCh = inputReader.read();
+          ch = (char)intCh;
+          if (isOrBar(ch))
+            terminating = true;
+          else if (terminating && isPound(ch))
+            done = true;
+          else
+            terminating = false;
+        }
+      } catch (IOException e) {
+        System.err.println("\n *** I/O error while reading a block comment.  Not terminated?");
+      }
+      return NIL;
+    }
+
+    else
+    {
+      System.err.println("\n *** unknown '#' construct.");
+      return NIL;
+    }
+  }
+
+
+  /**
+   * This library can't read backquotes yet.
+   */
+  public LispValue read_backquoted_list_token(PushbackReader stream)
+  {
+    System.err.println("\n *** Parser can't read backquoted lists yet.");
+    return NIL;
+  }
+
+	/**
+	 * Converts a string to a LISP value such as
+	 * NIL, T, an integer, a real number, a string or a symbol.
+	 */
+	public LispValue tokenToLispValue(String token)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.